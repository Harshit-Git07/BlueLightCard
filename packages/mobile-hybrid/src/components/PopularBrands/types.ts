export interface PopularBrandsProps {
  brands: Brand[];
  text?: string;
  onBrandItemClick?: (data: any) => void;
}

export interface Brand {
  id: number;
  imageSrc: string;
  altText: string;
}
