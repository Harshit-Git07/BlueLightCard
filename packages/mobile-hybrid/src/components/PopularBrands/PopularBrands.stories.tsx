import { Meta, StoryFn } from '@storybook/react';
import PopularBrands from './PopularBrands';
import brands from '@/modules/popularbrands/brands';

const componentMeta: Meta<typeof PopularBrands> = {
  title: 'PopularBrands',
  component: PopularBrands,
};

const DefaultTemplate: StoryFn<typeof PopularBrands> = (args) => <PopularBrands {...args} />;

export const Default = DefaultTemplate.bind({});

Default.args = {
  title: 'Popular Brands',
  brands,
};

export default componentMeta;
