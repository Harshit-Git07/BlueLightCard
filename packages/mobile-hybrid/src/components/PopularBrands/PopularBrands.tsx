import React, { FC } from 'react';
import { PopularBrandsProps } from './types';
import Image from '@/components/Image/Image';
import Heading from '../Heading/Heading';
import { cssUtil } from '@/utils/cssUtil';
import useDebounce from '@/hooks/useDebounce';

import { useAmplitude } from '@/hooks/useAmplitude';
import { Experiments } from '@/components/AmplitudeProvider/amplitudeKeys';

const { POPULAR_BRANDS_INLINE_LAYOUT } = Experiments;

const PopularBrands: FC<PopularBrandsProps> = ({
  rounded = true,
  title,
  brands,
  onBrandItemClick,
  onInteracted,
  text,
}) => {
  const throttle = useDebounce(() => {
    if (onInteracted) {
      onInteracted();
    }
  }, 750);

  const imageWrapperClass = cssUtil([
    'popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md',
    rounded ? 'rounded-full w-[74px]' : 'rounded w-[144px]',
  ]);

  const imageClass = cssUtil([
    'object-contain dark:border',
    rounded ? 'rounded-full' : 'object-cover rounded',
  ]);

  const { is } = useAmplitude();

  return (
    <div className="mb-6">
      <Heading title={title} />
      <p className="px-4 mb-4 dark:text-neutral-white">{text}</p>
      {is(POPULAR_BRANDS_INLINE_LAYOUT, 'treatment') ? (
        <div onScroll={throttle}>
          <div className="grid grid-cols-4 gap-4 px-4 mx-[-0.5rem]">
            {brands.map((brand, index) => {
              if (index > 7) return; // limit to 8 items

              return (
                <button
                  key={brand.id}
                  className={imageWrapperClass}
                  onClick={() => onBrandItemClick?.(brand.id)}
                >
                  <Image src={brand.imageSrc} alt={brand.brandName} className={imageClass} />
                </button>
              );
            })}
          </div>
        </div>
      ) : (
        <div className="ml-2 flex overflow-x-auto items-center h-[100px]" onScroll={throttle}>
          <div className="flex">
            {brands.map((brand) => (
              <button
                key={brand.id}
                className={imageWrapperClass}
                onClick={() => onBrandItemClick?.(brand.id)}
              >
                <Image src={brand.imageSrc} alt={brand.brandName} className={imageClass} />
              </button>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default PopularBrands;
