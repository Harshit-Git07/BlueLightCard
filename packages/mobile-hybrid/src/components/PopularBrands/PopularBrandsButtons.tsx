import React from 'react';
import { Brand } from './types';
import Image from '@/components/Image/Image';

interface Props {
  brands: Brand[];
  limit?: number;
  imageWrapperClass: string;
  imageClass: string;
  onButtonClick?: (data: any) => void;
}

const PopularBrandsButtons: React.FC<Props> = ({
  brands,
  limit,
  imageWrapperClass,
  imageClass,
  onButtonClick,
}: Props) => {
  return (
    <>
      {brands.map((brand: any, index: any) => {
        if (limit && index > limit) return; // limit to 8 items

        return (
          <button
            key={brand.id}
            className={imageWrapperClass}
            onClick={() => onButtonClick?.(brand.id)}
            aria-label={brand.brandName}
          >
            <Image src={brand.imageSrc} alt={brand.brandName} className={imageClass} />
          </button>
        );
      })}
    </>
  );
};

export default PopularBrandsButtons;
