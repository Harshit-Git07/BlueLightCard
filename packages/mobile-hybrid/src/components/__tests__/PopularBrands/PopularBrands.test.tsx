import React from 'react';
import { render, screen, fireEvent, within } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import PopularBrands from '@/components/PopularBrands/PopularBrands';

const brands = [
  {
    id: 22,
    imageSrc: '/images/image.jpg',
    brandName: 'Gym King',
  },
  {
    id: 23,
    imageSrc: '/images/image.jpg',
    brandName: 'M&S',
  },
  {
    id: 24,
    imageSrc: '/images/image.jpg',
    brandName: 'Asda',
  },
  {
    id: 25,
    imageSrc: '/images/image.jpg',
    brandName: 'Burger King',
  },
  {
    id: 26,
    imageSrc: '/images/image.jpg',
    brandName: 'Toyota',
  },
  {
    id: 27,
    imageSrc: '/images/image.jpg',
    brandName: 'McDonalds',
  },
  {
    id: 28,
    imageSrc: '/images/image.jpg',
    brandName: 'Superdry',
  },
  {
    id: 29,
    imageSrc: '/images/image.jpg',
    brandName: 'Sainsburys',
  },
  {
    id: 30,
    imageSrc: '/images/image.jpg',
    brandName: 'David Lloyd',
  },
];

// Mock the next/image component
jest.mock('next/image', () => {
  return function MockImage(props: any) {
    // eslint-disable-next-line @next/next/no-img-element, jsx-a11y/alt-text
    return <img {...props} />;
  };
});

jest.useFakeTimers(); // Use fake timers for testing

describe('PopularBrands', () => {
  const mockProps = {
    rounded: false,
    title: 'Popular brands test',
    brands: brands,
    text: 'Some of your favourite brands',
    onBrandItemClick: jest.fn(),
    onInteracted: jest.fn(),
  };

  it('renders the component', () => {
    const { container } = render(<PopularBrands {...mockProps} />);

    window.scrollY = 100;
    window.dispatchEvent(new Event('scroll'));

    jest.advanceTimersByTime(750);

    expect(screen.getByText('Popular brands test')).toBeInTheDocument();
    expect(screen.getByText('Some of your favourite brands')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('is rounded', () => {
    const { container } = render(<PopularBrands {...mockProps} rounded={true} />);
    const button = container.querySelectorAll('.popular-brands') as unknown as HTMLButtonElement[];

    const image = within(button[0]).getByAltText('Gym King');

    expect(image).toBeInTheDocument();
  });

  it('renders buttons with images', () => {
    const { getByAltText } = render(<PopularBrands {...mockProps} />);

    brands.forEach((brand) => {
      const image = getByAltText(brand.brandName) as HTMLImageElement;

      expect(image).toBeInTheDocument();
      expect(image.src).toContain('/images/image.jpg');
    });
  });

  it('calls the onBrandItemClick function when clicked', async () => {
    const { container } = render(<PopularBrands {...mockProps} />);
    const button = container.querySelectorAll('.popular-brands') as unknown as HTMLButtonElement[];

    fireEvent.click(button[0]);

    expect(mockProps.onBrandItemClick).toHaveBeenCalled();
  });

  it('renders when it has at least 8 items', () => {
    const { container } = render(<PopularBrands {...mockProps} />);
    const buttons = container.querySelectorAll('.popular-brands') as unknown as HTMLButtonElement[];

    expect(buttons[0]).toBeInTheDocument();

    expect(buttons).toHaveLength(9);
  });
});
