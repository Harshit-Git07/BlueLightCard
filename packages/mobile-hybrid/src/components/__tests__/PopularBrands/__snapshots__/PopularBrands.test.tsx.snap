// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopularBrands renders the component 1`] = `
<div>
  <div
    class="mb-6"
  >
    <div
      class="flex w-full content-center my-2 px-4"
    >
      <h2
        class="font-heading-font font-heading-font-weight text-heading-font tracking-heading-font leading-heading-font flex-1 text-heading-colour-light dark:text-heading-colour-dark"
      >
        Popular brands test
      </h2>
    </div>
    <p
      class="px-4 mb-4 dark:text-neutral-white"
    >
      Some of your favourite brands
    </p>
    <div
      class="ml-2 flex overflow-x-auto items-center h-[100px]"
    >
      <div
        class="flex"
      >
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="Gym King"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="M&S"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="Asda"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="Burger King"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="Toyota"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="McDonalds"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="Superdry"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="Sainsburys"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
        <button
          class="popular-brands mx-2 inline-block relative h-[74px] overflow-hidden shadow-md rounded w-[144px]"
        >
          <img
            alt="David Lloyd"
            blurdataurl="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mPs7+9bAQAFtwJW4mO4owAAAABJRU5ErkJggg=="
            class="object-contain dark:border object-cover rounded"
            placeholder="blur"
            src="/images/image.jpg"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
